### PR TITLE
FundingLocked improvements

### PR DESCRIPTION
**Pull request by halseth**
This PR contains several improvements to the way `fundingLocked` messages are sent and handled, to be able to continue the channel opening process in edge cases like restarts and disconnects.

This closes issue #303.

1. We now ignore a duplicate received `fundingLocked`.
2. In the case we start the `channelLink` and the number of updates recorded on that channel is 0, we resend the `fundingLocked` message. This is and intermediate step before merging #231.
3. In case the `fundingManager` gets restarted before the channel opening process is finished, we set up channel barriers again, such that `waitUntilChannelOpen` correctly will wait for it to finish.
4. Instead of sending the `fundingLocked` message synchronously to the peer, we add a listener to a new queue in the `server`, such that the fundingManager will get notified when the peer comes online. This is important since the peer might have gone offline while we were waiting for the funding tx to confirm, or if we just restarted `lnd`, the connection might not be established yet.

### CODE REVIEW

**Comment by halseth** [comment]
Moved resend logic of `fundingLocked` from `server` to `fundingManager`, now letting other subsystems register `on connected` listeners for peers. 

PTAL

File: lnwallet/channel.go [hunk 1]
```diff
@@ -3961,3 +3961,8 @@ func CreateCooperativeCloseTx(fundingTxIn *wire.TxIn,
 func (lc *LightningChannel) CalcFee(feeRate uint64) uint64 {
 	return (feeRate * uint64(commitWeight)) / 1000
 }
+
+// RemoteNextRevocation returns the channelState's RemoteNextRevocation.
+func (lc *LightningChannel) RemoteNextRevocation() *btcec.PublicKey {
+	return lc.channelState.RemoteNextRevocation
```
Comment (Roasbeef) [inline lnwallet/channel.go]: This method should be holding on to the mutex embedded within this struct. Otherwise, a race condition can arise between this method and `ReceiveRevocation` or `SignNextCommitment`.

File: fundingmanager.go [hunk 2]
```diff
@@ -420,10 +424,22 @@ func (f *fundingManager) Start() error {
...
+		// Set up the channel barriers again, to make sure
+		// waitUntilChannelOpen correctly waits until the opening
+		// process is completely over.
+		f.barrierMtx.Lock()
+		fndgLog.Tracef("Loading pending ChannelPoint(%v), "+
+			"creating chan barrier", channel.FundingOutpoint)
+		f.newChanBarriers[chanID] = make(chan struct{})
+		f.barrierMtx.Unlock()
+
+		// Set up a localDiscoverySignals to make sure we finish sending
```
Comment (Roasbeef) [inline fundingmanager.go]: +1 on the added comment!

File: fundingmanager.go [hunk 3]
```diff
@@ -1500,17 +1517,46 @@ func (f *fundingManager) processFundingLocked(msg *lnwire.FundingLocked,
 // handleFundingLocked finalizes the channel funding process and enables the
 // channel to enter normal operating mode.
 func (f *fundingManager) handleFundingLocked(fmsg *fundingLockedMsg) {
+	defer f.wg.Done()
+
+	// If we are currently in the process of handling a funding locked
```
Comment (Roasbeef) [inline fundingmanager.go]: Why is this necessary?

File: fundingmanager_test.go [hunk 4]
```diff
@@ -150,7 +147,8 @@ func createTestFundingManager(t *testing.T, pubKey *btcec.PublicKey,
 	tempTestDir string, hdSeed []byte, netParams *chaincfg.Params,
 	chainNotifier chainntnfs.ChainNotifier, estimator lnwallet.FeeEstimator,
 	sentMessages chan lnwire.Message, sentAnnouncements chan lnwire.Message,
-	publTxChan chan *wire.MsgTx, shutdownChan chan struct{}) (*fundingManager, error) {
+	publTxChan chan *wire.MsgTx, shutdownChan chan struct{},
```
Comment (Roasbeef) [inline fundingmanager_test.go]: At first glance, the amount of parameters passed into this function is a bit excessive. Instead, it seems possible that this function itself can instantiate all the items required to create a test funding manager _within_, and instead return a fully populated "test fixture" for the funding manager itself which can be manipulated as necessary within the tests.

File: fundingmanager_test.go [hunk 5]
```diff
@@ -347,20 +390,24 @@ func setupFundingManagers(t *testing.T, shutdownChannel chan struct{}) (*testNod
...
 	}
 
 	return alice, bob
 }
 
-func tearDownFundingManagers(t *testing.T, a, b *testNode, shutdownChannel chan struct{}) {
-	close(shutdownChannel)
+func tearDownFundingManagers(t *testing.T, a, b *testNode) {
+	close(a.shutdownChannel)
+	close(b.shutdownChannel)
```
Comment (Roasbeef) [inline fundingmanager_test.go]: Perhaps a wait group should be introduced to ensure that the to test nodes are _actually_ fully shutdown before proceeding?

File: fundingmanager_test.go [hunk 6]
```diff
@@ -825,10 +927,23 @@ func TestFundingManagerRestartBehavior(t *testing.T) {
...
-		t.Fatalf("expected fundingLocked sent from Alice, "+
-			"instead got %T", fundingLockedAlice)
+	var aliceMsg lnwire.Message
+	select {
+	case aliceMsg = <-alice.msgChan:
+	case <-time.After(time.Second * 5):
+		t.Fatalf("alice did not send fundingLocked")
+	}
+
+	fundingLockedAlice, ok := aliceMsg.(*lnwire.FundingLocked)
```
Comment (Roasbeef) [inline fundingmanager_test.go]: I count at least three repetitions of this identical code block. The check should be factor out into something like an `assertMessageIsFundingLocked` (or something along those lines).

File: fundingmanager_test.go [hunk 7]
```diff
@@ -953,15 +1068,45 @@ func TestFundingManagerRestartBehavior(t *testing.T) {
 		t.Fatalf("expected to not find channel state, but got: %v", state)
 	}
 
+	// Exchange the fundingLocked messages.
```
Comment (Roasbeef) [inline fundingmanager_test.go]: An identical code block is copy-pasta'd verbatim above in `TestFundingManagerNormalWorkflow`. The logic should be factor out into something along the lines of `assertFundingFlowResolution`.

File: fundingmanager_test.go [hunk 8]
```diff
@@ -1016,3 +1161,805 @@ func TestFundingManagerFundingTimeout(t *testing.T) {
...
+	select {
+	case aliceMsg = <-alice.msgChan:
+	case <-time.After(time.Second * 5):
+		t.Fatalf("alice did not send fundingLocked")
+	}
+
+	_, ok := aliceMsg.(*lnwire.FundingLocked)
+	if !ok {
+		errorMsg, gotError := aliceMsg.(*lnwire.Error)
+		if gotError {
```
Comment (Roasbeef) [inline fundingmanager_test.go]: Another instance of this repeated code fragment.

File: fundingmanager_test.go [hunk 9]
```diff
@@ -1016,3 +1161,805 @@ func TestFundingManagerFundingTimeout(t *testing.T) {
...
+	state, _, err = bob.fundingMgr.getChannelOpeningState(fundingOutPoint)
+	if err != nil {
+		t.Fatalf("unable to get channel state: %v", err)
+	}
+
+	if state != fundingLockedSent {
+		t.Fatalf("expected state to be fundingLockedSent, was %v", state)
+	}
+
+	// After the FundingLocked message is sent, the channel will be announced.
```
Comment (Roasbeef) [inline fundingmanager_test.go]: This code fragment is repeated in just about all the unit tests for this sub-system. Factoring it out to a single test helper func will reduce the size of this diff (and this file) by a good bit.

File: fundingmanager_test.go [hunk 10]
```diff
@@ -1016,3 +1161,805 @@ func TestFundingManagerFundingTimeout(t *testing.T) {
...
+	case bobMsg = <-bob.msgChan:
+	case <-time.After(time.Second * 5):
+		t.Fatalf("bob did not send fundingLocked")
+	}
+
+	fundingLockedBob, ok := bobMsg.(*lnwire.FundingLocked)
+	if !ok {
+		errorMsg, gotError := bobMsg.(*lnwire.Error)
+		if gotError {
+			t.Fatalf("expected FundingLocked to be sent "+
```
Comment (Roasbeef) [inline fundingmanager_test.go]: Similar comment about this duplicated code block to those above (won't comment again to avoid cluttering the review comments).

File: fundingmanager_test.go [hunk 11]
```diff
@@ -1016,3 +1161,805 @@ func TestFundingManagerFundingTimeout(t *testing.T) {
...
+
+	if state != fundingLockedSent {
+		t.Fatalf("expected state to be fundingLockedSent, was %v", state)
+	}
+
+	// After the FundingLocked message is sent, the channel will be announced.
+	// A chanAnnouncement consists of three distinct messages:
+	//	1) ChannelAnnouncement
+	//	2) ChannelUpdate
+	//	3) AnnounceSignatures
```
Comment (Roasbeef) [inline fundingmanager_test.go]: Should be factored into the function proposed above.

File: htlcswitch/switch.go [hunk 12]
```diff
@@ -894,9 +894,12 @@ func (s *Switch) addLink(link ChannelLink) error {
 		return err
 	}
 
-	log.Infof("Added channel link with chan_id=%v, short_chan_id=(%v), "+
-		"bandwidth=%v", link.ChanID(), spew.Sdump(link.ShortChanID()),
-		link.Bandwidth())
+	// TODO(halseth): remove this after roasbeef's commits.
```
Comment (Roasbeef) [inline htlcswitch/switch.go]: After my latest commits within the `htlcswitch`, it should be possible to remove this `go` statement, and not re-introduce the deadlock case.

File: htlcswitch/test_utils.go [hunk 13]
```diff
@@ -449,7 +449,7 @@ func (n *threeHopNetwork) makePayment(sendingPeer, receivingPeer Peer,
 	select {
 	case err := <-errChan:
 		return invoice, err
-	case <-time.After(5 * time.Minute):
+	case <-time.After(15 * time.Second):
```
Comment (Roasbeef) [inline htlcswitch/test_utils.go]: When running with the `-race` flag on Travis, I've seen this case take up to `1 minute` to complete, which is why I had the timeout set to 5. Normally it completes very quickly, but Travis is a different beast...

File: server.go [hunk 14]
```diff
@@ -863,6 +866,33 @@ func (s *server) SendToPeer(target *btcec.PublicKey,
...
+	defer s.mu.Unlock()
+
+	// Compute the target peer's identifier.
+	pubStr := string(peer.SerializeCompressed())
+
+	// Check if peer is connected.
+	_, ok := s.peersByPub[pubStr]
+	if ok {
+		// Connected, can return early.
+		srvrLog.Debugf("notifying that peer %v is online", pubStr)
```
Comment (Roasbeef) [inline server.go]: Minor nit: logging statement should start with a capital letter.

File: server.go [hunk 15]
```diff
@@ -1275,6 +1305,11 @@ func (s *server) addPeer(p *peer) {
 	// channel router so we can synchronize our view of the channel graph
 	// with this new peer.
 	go s.authGossiper.SynchronizeNode(p.addr.IdentityKey)
+
+	// Check if there are listeners waiting for this peer to come online.
+	for _, con := range s.peerConnectedListeners[pubStr] {
+		close(con)
```
Comment (Roasbeef) [inline server.go]: I believe this may return an `nil` value if the key has not yet been populated. As a result, this may result in a panic due to attempting to close a `nil` channel.

File: fundingmanager.go [hunk 16]
```diff
@@ -1437,18 +1443,44 @@ func (f *fundingManager) sendFundingLockedAndAnnounceChannel(
...
+	// where we might try to resend the fundingLocked message before the
+	// server has had the time to connect to the peer.
 	err = f.cfg.SendToPeer(completeChan.IdentityPub, fundingLockedMsg)
 	if err != nil {
-		fndgLog.Errorf("unable to send fundingLocked to peer: %v", err)
-		return
+		fndgLog.Warnf("unable to send fundingLocked to peer: %v. "+
+			"Will retry when online", err)
+
+		connected := make(chan struct{})
```
Comment (Roasbeef) [inline fundingmanager.go]: In order to handle a possible flapping case, this should be performed within a loop, something along the lines of: 
```go
for {
     // attempt to send

     // if send succeeds, then break
  
     // re-register the ntfn
}
```

File: fundingmanager_test.go [hunk 17]
```diff
@@ -967,139 +943,180 @@ func TestFundingManagerRestartBehavior(t *testing.T) {
...
-		case *lnwire.AnnounceSignatures:
-			gotAnnounceSignatures = true
-		case *lnwire.NodeAnnouncement:
-			gotNodeAnnouncement = true
-		}
-	}
-
-	if !gotChannelAnnouncement {
-		t.Fatalf("did not get ChannelAnnouncement from Alice after restart")
-	}
```
Comment (Roasbeef) [inline fundingmanager_test.go]: Ah, going through the commits in order so I see this refactoring now. If these changes fully addresses my prior commits, then great, otherwise please ensure that duplication has been eliminated.

File: fundingmanager.go (hunk 3 above)
Comment (halseth) [reply fundingmanager.go]: This is for the case where the peer is sending two `fundingLocked` messages rapidly after one another. In this case we couldn't check the `RemoteNextRevocation` of the channel, since we had no guarantees we were done handling the first message. This is special for `handleFundingLocked`, since it is run as a goroutine.

File: fundingmanager_test.go (hunk 5 above)
Comment (halseth) [reply fundingmanager_test.go]: I don't think it is strictly necessary. The shutdownChannel is used to make sure mocked methods like `SendToPeer` returns if they are waiting for a channel send at the moment of shutdown. Immediately after closing the shutdownChannel, we call `fundingMngr.Stop()`, which will wait for all the go routines internal to fundingManager to stop. Closing the shutdown channel makes sure this `Stop()` never returns.

File: fundingmanager_test.go (hunk 17 above)
Comment (halseth) [reply fundingmanager_test.go]: Yep, in the last commit I got rid of as much duped code as I could :)

File: htlcswitch/test_utils.go (hunk 13 above)
Comment (halseth) [reply htlcswitch/test_utils.go]: Do'h! reverting then 🤠

File: server.go (hunk 15 above)
Comment (halseth) [reply server.go]: If not populated, a nil value will be returned from the map, but I think ranging over nil slices is safe. I created an example here:
https://play.golang.org/p/S1Rhxjk2Ha

File: fundingmanager_test.go (hunk 4 above)
Comment (halseth) [reply fundingmanager_test.go]: Totally agree. Refactored :)

**Comment by halseth** [comment]
Addressed review comments. PTAL.

File: fundingmanager_test.go [hunk 18]
```diff
@@ -702,31 +683,117 @@ func TestFundingManagerNormalWorkflow(t *testing.T) {
...
+	// Exchange the fundingLocked messages.
+	alice.fundingMgr.processFundingLocked(fundingLockedBob, bobAddr)
+	bob.fundingMgr.processFundingLocked(fundingLockedAlice, aliceAddr)
+
+	// Check that they notify the breach arbiter and peer about the new
+	// channel.
+	assertHandleFundingLocked(t, alice, bob)
+}
+
+func TestFundingManagerRestartBehavior(t *testing.T) {
```
Comment (cfromknecht) [inline fundingmanager_test.go]: A++ on these test cases

File: fundingmanager.go [hunk 19]
```diff
@@ -1420,18 +1443,45 @@ func (f *fundingManager) sendFundingLockedAndAnnounceChannel(
...
+	// message. This is special for fundingLocked, since failing to send any
+	// of the previous messages in the funding flow just cancels the flow.
+	// But now the funding transaction is confirmed, the channel is open
+	// and we have to make sure the peer gets the fundingLocked message when
+	// it comes back online. This is also crucial during restart of lnd,
+	// where we might try to resend the fundingLocked message before the
+	// server has had the time to connect to the peer. We keep trying to
+	// send fundingLocked until we succeed, or the fundingManager is shut
+	// down.
+	for {
```
Comment (cfromknecht) [inline fundingmanager.go]: Really like the design of this retry logic, looks like it will be very reliable 🚀 👏

Comment (halseth) [reply fundingmanager.go]: 🚀

File: server.go (hunk 15 above)
Comment (halseth) [reply server.go]: whoa, you are totally right! I pushed a fix, PTAL. Good catch!

Comment (cfromknecht) [reply server.go]: LGTM 😎

File: fundingmanager.go [hunk 20]
```diff
@@ -1420,18 +1443,45 @@ func (f *fundingManager) sendFundingLockedAndAnnounceChannel(
...
+		err = f.cfg.SendToPeer(completeChan.IdentityPub,
+			fundingLockedMsg)
+		if err == nil {
+			// Sending succeeded, we can break out and continue
+			// the funding flow.
+			break
+		}
+
+		fndgLog.Warnf("unable to send fundingLocked to peer: "+
+			"%v. Will retry when online", err)
```
Comment (Roasbeef) [inline fundingmanager.go]: Should be logging the serialized compressed version of the peer here instead: 
```go
fndgLog.Warnf("unable to send fundingLocked to peer: "+
"%x. Will retry when online", completeChan.IdentityPub.SerializeCompressed())
```

File: fundingmanager.go [hunk 21]
```diff
@@ -1554,9 +1640,13 @@ func (f *fundingManager) handleFundingLocked(fmsg *fundingLockedMsg) {
 		// that commitment related modifications to this channel can
 		// now proceed.
 		f.barrierMtx.Lock()
-		fndgLog.Tracef("Closing chan barrier for ChanID(%v)", chanID)
-		close(f.newChanBarriers[chanID])
-		delete(f.newChanBarriers, chanID)
+		chanBarrier, ok := f.newChanBarriers[chanID]
```
Comment (Roasbeef) [inline fundingmanager.go]: We're missing a conditional select above, on line 1634. This existed _before_ this PR, but we should fix it now while we're in the area:
```go
select {
    case f.cfg.ArbiterChan <- channel:
    case <-f.quit:
        return
}
```

File: fundingmanager.go (hunk 20 above)
Comment (halseth) [reply fundingmanager.go]: Added logging of the peer ID. I think we still want to keep logging the error though.

File: fundingmanager.go (hunk 21 above)
Comment (halseth) [reply fundingmanager.go]: Added this, and also a similar select for `peer.newChannels <- newChanMsg` below.

**Comment by halseth** [comment]
Rebased and addressed comments 😎 

PTAL.
